axlearn/common/config.py

```diff
@@ -394,6 +394,66 @@ def set(self, **kwargs):
             setattr(self, k, v)
         return self
 
+    class TraverseResult(NamedTuple):
+        """Result of a recurisve traverse in a nested ConfigBase."""
+
+        # The parent that contains the reulting key.
+        parent: _ConfigBase
+        # The key string.
+        key: str
+
+    def recursive_traverse(self, key_path: Sequence[str]) -> tuple[Any, str]:
+        """Recursively traverse the config to find the target key.
+
+        Args:
+            key_path: A sequence of keys for indexing.
+
+        Raises:
+            ValueError: A key in key_path is not found.
+
+        Returns:
+            A tuple containing the parent object and the target key.
+        """
+        target_key = key_path[0]
+        if target_key not in self:
+            raise ValueError(f"{target_key} is not found in {self}.")
+
+        if len(key_path) == 1:
+            # Completed traversal
+            return self.TraverseResult(parent=self, key=key_path[0])
+
+        # Continue searching recursively
+        value = getattr(self, target_key)
+        return value.recursive_traverse(key_path[1:])
+
+    def get_recursively(self, key_path: Sequence[str]) -> Any:
+        """Recursively find the target key in the config and return its value.
+
+        Args:
+            key_path: A sequence of keys for indexing to get the target value.
+
+        Raises:
+            ValueError: A key in key_path is not found.
+
+        Returns:
+            value at the key_path.
+        """
+        traverse_result = self.recursive_traverse(key_path)
+        return getattr(traverse_result.parent, traverse_result.key)
+
+    def set_recursively(self, key_path: Sequence[str], new_value: Any):
+        """Recursively find the target key in the config and set its value.
+
+        Args:
+            key_path: A sequence of keys for indexing to set the target value.
+            new_value: New value to replace the target value.
+
+        Raises:
+            ValueError: A key in key_path is not found.
+        """
+        traverse_result = self.recursive_traverse(key_path)
```

Comment:
Can we do something like:
```suggestion
        if not path:
            raise ValueError(...)
        parent = self.get_recursively(path[:-1])
```

Reply:
I removed the `recursive_traverse` method and added a value error if path is empty. I am still using path[1:] instead path[:-1] since we are traversing parents first.